[[ Bug 17810 ]] Non-LF line endings in script only stacks

bwmilby opened the pull request:
No code changes were necessary for LC9 to support reading files with the various line endings.  Separate PR outstanding for BOM handling.

Added 2 bool flags to the stack struct to retain state of line endings used when importing a script only stack.  LF is set to true upon object creation as the default for new script only stacks.  (Could use an #if defined to set CRLF for Windows if desired)
Added methods to get/set the line ending values on a stack object.

[Considered using MCSOutputLineEndings enum (osspec.h), but Monte mentioned it would need to get moved to foundation first.  Changing the code would not be very difficult though.  kMCSOutputLineEndingsNative is not really correct for Mac since __CR__ is still used and I understand can't be changed right now.  If we changed to this enum, then the default would need to be LF except for Windows.]

When the stack object is created for a script only stack, the raw script is checked for CR and LF line endings.  If neither found, then LF is selected.
MCStringConvertLineEndingsFromLiveCode was overloaded with a version that accepts the 2 bool values for CR and LF endings.  The save stack call was modified to use this version with the line ending values from the stack object.

bwmilby committed:
[[ Bug 17810 ]] Non-LF line endings in script only stacks

No code changes were necessary for LC9 to support reading files with the various line endings.  Separate PR outstanding for BOM handling.
Added 2 bool flags to the stack struct to retain state of line endings used when importing a script only stack.  LF is set to true upon object creation as the default for new script only stacks.  (Could use an #if defined to set CRLF for Windows if desired)
Added methods to get/set the line ending values on a stack object.
[Considered using MCSOutputLineEndings enum (osspec.h), but Monte mentioned it would need to get moved to foundation first.  Changing the code would not be very difficult though.  kMCSOutputLineEndingsNative is not really correct for Mac since __CR__ is still used and I understand can't be changed right now.  If we changed to this enum, then the default would need to be LF except for Windows.]
When the stack object is created for a script only stack, the raw script is checked for CR and LF line endings.  If neither found, then LF is selected.
MCStringConvertLineEndingsFromLiveCode was overloaded with a version that accepts the 2 bool values for CR and LF endings.  The save stack call was modified to use this version with the line ending values from the stack object.

runrevmark commented:
@bwmilby Thanks for this :)

It looks like it is getting there, however I wonder if we do need a libfoundation MCString function here to do the heavy lifting. Indeed, we need to account for three types of line-ending UNIX (LF), Legacy Mac (CR) and Windows (CR LF). We could add a 'NormalizeLineEndings' libfoundation function which normalizes to a specific style, and returns the *first* style it encounters whilst doing so - then the object property can be a single field with the line ending style.

```
enum MCStringLineEndingStyle
{
  kMCStringLineEndingStyleLF,
  kMCStringLineEndingStyleCR,
  kMCStringLineEndingStyleCRLF
}

bool MCStringNormalizeLineEndings(MCStringRef p_input, MCStringLineEndingStyle p_to_style, MCStringRef& r_output, MCStringLineEndingStyle* r_original_style);
```

Note: Using a ptr rather than a reference as the last parameter means it can be optionally provided. In some cases you might not care what the original style was!

bwmilby commented:
I can add the enum and function.  The existing code will preserve all 3 combinations (for a file that doesn't use mixed endings), but an enum will make it more clear.

@runrevmark Since this is going to move to libfoundation, should I also go ahead and update the other uses of the line ending conversion calls (there were not very many that I found)?

bwmilby committed:
Added suggested MCStringLineEndingStyle enum and MCStringNormalizeLineEndings function.  Did not remove the other added code yet nor convert other line ending conversion calls to the new version yet.

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -6963,6 +6963,60 @@ bool MCStringGetNumericValue(MCStringRef self, double &r_value)
+                                  MCStringRef& r_output, 
+                                  MCStringLineEndingStyle* r_original_style)
+{
+    uindex_t t_firstLF = 0;
```
Please use snake-case for vars - i.e. t_first_lf

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -6963,6 +6963,60 @@ bool MCStringGetNumericValue(MCStringRef self, double &r_value)
+    MCStringLineEndingStyle t_original_style = kMCStringLineEndingStyleLF;
+    MCStringRef t_mutable_input;
+    
+    if (MCStringFirstIndexOfChar(p_input, '\r', 0, kMCStringOptionCompareExact, t_firstCR))
```
I suggest variable definitions should be just before the block which first uses them, with ones which are needed outside the block first:
```
MCStringLineEndingStyle t_original_style...
uindex_t t_first_lf ...
uindex_t t_first_cr
```

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -6963,6 +6963,60 @@ bool MCStringGetNumericValue(MCStringRef self, double &r_value)
+    }
+    
+    // normalize input to LF line endings first
+    /* UNCHECKED */ MCStringMutableCopy(p_input, t_mutable_input);
```
Move the declaration of t_mutable_input to here and use an auto-class:
```
MCAutoStringRef t_mutable_input;
```
You can then easily make all the calls checked:
```
if (!MCStringMutableCopy(p_input, &t_mutable_input))
{
  return false;
}
```

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -6963,6 +6963,60 @@ bool MCStringGetNumericValue(MCStringRef self, double &r_value)
+    
+    // normalize input to LF line endings first
+    /* UNCHECKED */ MCStringMutableCopy(p_input, t_mutable_input);
+    /* UNCHECKED */ MCStringFindAndReplace(t_mutable_input, MCSTR("\r\n"), 
```
Normalizing to LF only requires two replacements:
```
replace crlf with lf in tString
replace cr with lf in tString
```
In particular, LF-CR is a degenerate case, which can be seen as a mixed line-ending file using LF and CR.

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -6963,6 +6963,60 @@ bool MCStringGetNumericValue(MCStringRef self, double &r_value)
+                        kMCStringOptionCompareExact);
+    
+    // now convert the line endings to the proper version
+    if (p_to_style == kMCStringLineEndingStyleCR)
```
Uses braces around if blocks - regardless of number of lines. It reduces potential code change in the future (if a block requires addition of an extra statement) and avoids dangling else ambiguities.
```
if (p_to_style ...)
{ 
  ...
}
else
{
  ...
}
```

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -6963,6 +6963,60 @@ bool MCStringGetNumericValue(MCStringRef self, double &r_value)
+        /* UNCHECKED */ MCStringFindAndReplace(t_mutable_input, MCSTR("\n"), 
+                                               MCSTR("\r\n"), kMCStringOptionCompareExact);
+    
+    if (r_original_style != nullptr) *r_original_style = t_original_style;
```
Don't use single-line ifs at all:
```
if (r_original_style != nullptr)
{
  *r_original_style = t_original_style;
}
```

runrevmark commented on libfoundation/include/foundation.h:
```diff
@@ -2723,6 +2723,19 @@ MC_DLLEXPORT bool MCStringNormalizedCopyNFKD(MCStringRef, MCStringRef&);
+    kMCStringLineEndingStyleCRLF
+};
+
+MC_DLLEXPORT bool MCStringNormalizeLineEndings(
```
Generally for new libfoundation functions the style is:
```
MC_DLLEXPORT bool
MCStringNormalizeLineEndings(MCStringRef p_input,
                             MCStrignLineEndingStyle... ,
```

runrevmark replied:
We should add a descriptive comment just before the function definition:
```
// Converts LF, CR, CRLF line endings in p_input to p_to_style line endings and
// places the result into r_output. If p_normalize_ps_ls is true, then it also converts
// PS to p_to_style, and LS to VTAB.
```

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -6963,6 +6963,60 @@ bool MCStringGetNumericValue(MCStringRef self, double &r_value)
 }
 
+MC_DLLEXPORT
+bool MCStringNormalizeLineEndings(MCStringRef p_input, 
```
```
MC_DLLEXPORT bool
```

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -6963,6 +6963,60 @@ bool MCStringGetNumericValue(MCStringRef self, double &r_value)
+                                               MCSTR("\r\n"), kMCStringOptionCompareExact);
+    
+    if (r_original_style != nullptr) *r_original_style = t_original_style;
+    /* UNCHECKED */ MCStringCopyAndRelease(t_mutable_input, r_output);
```
With auto containers this becomes:
```
if (!t_mutable_input.MakeImmutable())
{
  return false;
}

r_output = t_mutable_input.Take();

return true;
```

bwmilby replied:
Moved the r_original_style below the possible failure point to avoid setting the value if not successful.

runrevmark commented on engine/src/stack.h:
```diff
@@ -300,6 +300,9 @@ class MCStack : public MCObject, public MCMixinObjectHandle<MCStack>
     bool m_is_script_only : 1;
+    
+    // BWM-2017-08-16: [[ Bug 17810 ]] Line endings for imported script-only-stack.
+    MCStringLineEndingStyle m_line_encoding_style;
```
You can use
```
MCStringLineEndingStyle m_line_ending_style : 2;
```
This won't break the bitfield, and is a little more economical on space.

montegoulding replied:
@runrevmark Hmm... I thought we might be discouraging bitfields these days https://github.com/livecode/livecode/pull/5259

montegoulding replied:
@peter-b 's comments are a little above my head there so I did some googling. I have found a number of suggestions not to use bitfields. Interestingly Mozilla's C++ portability guide recommends _only_ using them if they are all the same type and not to use them with enums. So we break both these rules here. https://developer.mozilla.org/en-US/docs/Mozilla/C++_Portability_Guide

bwmilby replied:
I made the change in what I'm about to upload, but it is easy to change back.

montegoulding replied:
@bwmilby my comments are more about educating myself. I'm sure I'll be wrong! Whatever @runrevmark says goes ;-)

runrevmark commented on engine/src/stack.h:
```diff
@@ -947,6 +950,10 @@ class MCStack : public MCObject, public MCMixinObjectHandle<MCStack>
     
+    // BWM-2017-08-16: [[ Bug 17810 ]] Get/set line endings for imported script-only-stack.
+    MCStringLineEndingStyle getlineencodingstyle(void) const { return m_line_encoding_style; }
+    void setlineencodingstyle(MCStringLineEndingStyle p_line_encoding_style);
```
It is probably worth putting the set definition inline here as well - as it doesn't do anything significant beyond update an internal var.

runrevmark commented on engine/src/dispatch.cpp:
```diff
@@ -1130,14 +1138,10 @@ IO_stat MCDispatch::dosavescriptonlystack(MCStack *sptr, const MCStringRef p_fna
-#else
-		t_converted = *t_script_body;
-#endif
+        MCStringNormalizeLineEndings(*t_script_body, sptr -> getlineencodingstyle(),
```
I suggest making the middle arguments go on their own line here. There are cases where it makes sense to have multiples ones on the same line (e.g. x, y) when they are related; otherwise separate lines helps readability for long argument lists.

runrevmark commented on engine/src/foundation-legacy.cpp:
```diff
@@ -314,6 +314,34 @@ bool MCCStringFromUnicode(const unichar_t *p_unicode_string, char*& r_string)
 
+// BWM-2017-08-16: [[ Bug 17810 ]] Restore line endings of script-only-stack to
+// match how they were when originally imported.
+bool MCStringConvertLineEndingsFromLiveCode(MCStringRef p_input, 
```
Do we still need this function?

bwmilby replied:
Not needed.

runrevmark commented on libfoundation/include/foundation.h:
```diff
@@ -2723,6 +2723,19 @@ MC_DLLEXPORT bool MCStringNormalizedCopyNFKD(MCStringRef, MCStringRef&);
+MC_DLLEXPORT bool MCStringNormalizeLineEndings(
+                            MCStringRef p_input, 
+                            MCStringLineEndingStyle p_to_style, 
+                            MCStringRef& r_output, 
```
I think this function actually (annoyingly) needs another parameter `bool p_normalize_ps_ls`. We want to do this when importing text into LiveCode, but not when exporting outside of LiveCode (otherwise you can't construct strings which contains PS / LS to export without having to do everything yourself - it's an edge case, but it is how the previous functions worked).

runrevmark commented:
@bmwilby: Thanks for the update - this is looking good! Your original code didn't handle the case of `LF ... CR` which I think would have ended up with it using CR-LF endings - however, I notice you've taken this into account in the revised version :)

I've added some comments (mostly stylistic) - however, can you look at refactoring the use of the other MCStringConvertLineEndings functions in foundation-legacy to using the new libfoundation API? I think with the extra (proposed) boolean parameter we can eliminate them.

runrevmark commented on engine/src/dispatch.cpp:
```diff
@@ -673,10 +673,14 @@ static MCStack* script_only_stack_from_bytes(uint8_t *p_bytes,
                                              MCStringEncoding p_encoding)
 {
     MCAutoStringRef t_raw_script_string, t_LC_script_string;
```
Snake-case var names, please - `t_lc_script_string` (or `t_lcs_string`)

runrevmark commented on engine/src/dispatch.cpp:
```diff
@@ -1119,6 +1126,7 @@ IO_stat MCDispatch::dosavescriptonlystack(MCStack *sptr, const MCStringRef p_fna
     
     // Compute the body of the script file.
 	MCAutoStringRef t_converted;
+    //MCStringLineEndingStyle t_line_encoding_style;
```
No need to keep this commented out line around

runrevmark commented on engine/src/foundation-legacy.h:
```diff
@@ -151,6 +151,7 @@ int UnicodeToUTF8(const uint16_t *lpSrcStr, int cchSrc, char *lpDestStr, int cch
 ////////////////////////////////////////////////////////////////////////////////
 
+bool MCStringConvertLineEndingsFromLiveCode(MCStringRef p_input, bool p_use_LF_line_endings, bool p_use_CR_line_endings, MCStringRef& r_output);
 bool MCStringConvertLineEndingsFromLiveCode(MCStringRef input, MCStringRef& r_output);
```
As mentioned - I wonder if we need any of these functions anywhere - I *think* they can all be replaced with MCStringNormalizeLineEndings after the extra bool parameter controlling lsps conversion is added.

bwmilby replied:
I would agree.  Just wanted to make sure I was on the right path before making those changes.  It is not very many.

runrevmark replied:
@bwmilby Ah! You mentioned that in your comment you made yday - so, in answer to that, yes... Unifying them makes sense :)

bwmilby commented:
I'll make all of the style changes today and refactor out the old calls.  For the legacy calls that would have made the PS/LS conversions, I'll set the flag to true.

bwmilby committed:
Migrated all 7 additional calls to the MCStringNormalizeLineEndings function.
Existing calls to the "to" version retained the PS/LS conversion.
Existing "from" versions were configured to use CRLF for Windows and LF otherwise.
Script only stacks will retain line endings that they originate in.  Default for new files is LF.
Made the requested style/format changes.  Removed LFCR detection (treated as LF, but will yield double spaced lines).
Compiled and verified on my Sierra system to open/save CR and CRLF files and retain line endings.

bwmilby commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -6963,6 +6963,109 @@ bool MCStringGetNumericValue(MCStringRef self, double &r_value)
+                                kMCStringOptionCompareExact))
+    {
+        return false;
+    }
```
To avoid double spaced files, could add:
`if (t_first_lf + 1 == t_first_cr) { /* convert LFCR to LF */ }`
This would maintain current behavior.

runrevmark replied:
I noticed the original code did that - but I don't know why - I think it might be just a case of 'over-engineering' there (i.e. not realizing CRLF->LF and CR->LF is enough to do the conversion).

LFCR isn't a valid line-ending, so wouldn't get generated by any text editor that exists - it could only come about (I think) if it were a file being edited in a line-ending preserving text editor which default to using CR. In thi s case it would actually be meant to be LF CR - i.e. two returns.

If I'm missing something here, please let me know!

bwmilby replied:
Don't think you are missing anything.  Just making sure the original code wasn't there for some special reason.

runrevmark commented on engine/src/stack.h:
```diff
@@ -300,6 +300,9 @@ class MCStack : public MCObject, public MCMixinObjectHandle<MCStack>
     bool m_is_script_only : 1;
+    
+    // BWM-2017-08-16: [[ Bug 17810 ]] Line endings for imported script-only-stack.
+    MCStringLineEndingStyle m_line_encoding_style : 2;
```
So, on the bitfield thing - I had forgotten that particular 'discussion' ;)

Bitfields are more than portable as long as you use them with care and don't use them in structs which are directly serialized... However *care* is needed.

The reason I suggested making this `: 2` was that it is between two other bitfields, so for consistency it seemed to make sense. However, looking at the entire MCStack header then it looks like it could save maybe 32 bytes or so by re-arrangement of the fields, but hardly essential (given the overhead of a stack in general, is quite large!).

It would have been fine if I had remembered that the signedness of bitfields vary - so this should be `: 3` to ensure that the top-bit is free (rule is always use 1 bit more than the number of bits needed to encode an enum's maximum value - then you don't get bit by the sign problem).

We perhaps need to update the style guide with better rules about field layout in classes and maybe still allow bitfields as long as they follow appropriate rules. Using a bitfield is pointless if there are less than 8 (8-bit or less sized) fields in a single struct/class, if there are many though they can make layouts more memory efficient; on the other hand, you can't default init them (i.e. bool m_is_script_only = false -- works; bool m_is_script_only : 1 = false -- doesn't) which means you have to have a constructor in a class which has bitfields. Swings, roundabouts!

runrevmark commented on engine/src/exec-files.cpp:
```diff
@@ -1953,8 +1953,19 @@ static void MCFilesReadComplete(MCExecContext& ctxt, MCValueRef p_output, IO_sta
+                                              &t_output, 
+                                              nullptr))
+            {
+                // handle error condition?
```
When a foundation function returns false then it will have set a libfoundation error at some point (almost all foundation functions only fail due to out-of-memory); so for normal Exec type functions, you would do `{ ctxt.Throw(); return }`. However, that depends on whether the caller is expecting an error from this function - which it doesn't look like it is. In this case, you'd probably be fine just using the same result as IO_ERROR.

Checking return values on foundation functions is more about cleanliness than anything else - they usually only fail due to OOM, which is generally uncatchable anyway (due to over-commit policies on modern OSes).

runrevmark commented on engine/src/exec-files.cpp:
```diff
@@ -2510,7 +2532,18 @@ void MCFilesExecWriteToProcess(MCExecContext& ctxt, MCNameRef p_process, MCStrin
 		MCStringRef t_text_data;
-		/* UNCHECKED */ MCStringConvertLineEndingsFromLiveCode(p_data, t_text_data);
+        if (!MCStringNormalizeLineEndings(p_data, 
+#ifdef __CRLF__
```
The original ConvertLineEndinsFromLiveCode also checked for __CR__ and used CR line-endings in that case (this is the default Mac behavior).

runrevmark replied:
See comment on PR as a whole about a better way to deal with this!

runrevmark commented on engine/src/sysspec.cpp:
```diff
@@ -1564,7 +1568,15 @@ bool MCS_savetextfile(MCStringRef p_filename, MCStringRef p_string)
     MCAutoStringRef t_converted;
     if (t_success)
-        t_success = MCStringConvertLineEndingsFromLiveCode(p_string, &t_converted);
+        t_success = MCStringNormalizeLineEndings(p_string, 
```
Need to handle __CR__ here too.

runrevmark replied:
See comment on PR as a whole about a better way to deal with this!

runrevmark commented:
@bwmilby This is looking really good - thanks for making the changes!

Having looked at the code after my suggested changes, I have a couple of other suggestions which will make the code much clearer.

The first is the handling of the 'default line ending' - currently we are using (as we did before) ifdef/elsif type arrangements to select the appropriate enum value. There's a better way to do this though - in libfoundation:
```
enum MCStringLineEndingStyle
{
  ... LF, CR, CRLF values ...
#if def(__CRLF__)
  kMCStringLineEndingStyleLegacyNative = kMCStringLineEndingStyleCRLF,
#elsif defined(__CR__)
  kMCStringLineEndingStyleLegacyNative = kMCStringLineEndingStyleCR,
#elif defined(__LF__)
  kMCStringLineEndingStyleLegacyNative = kMCStringLineEndingStyleLF,
#else
#error Unknown default line ending style (no __CRLF__, __CR__, __LF__ definition)
#endif
};
```

The second is the bool option we added to the foundation Normalize routine. I forgot to think in terms of 'readability' of the code - using a bool here is entirely opaque as there's no way of knowing what it means by reading the code. Instead, I should have suggested an options parameter:
```
typedef uint32_t MCStringLineEndingOptions;
enum
{
  /* Normalize any occurrence of PS (paragraph separator) to the specified line-ending style */
  kMCStringLineEndingOptionNormalizePSToLineEnding = 1 << 0,
  /* Normalize any occurrence of LS (line separator) to VT (vertical tab) */
  kMCStringLineEndingOptionNormalizeLSToVT = 1 << 1,
};

bool MCStringNormalizeLineEndings(MCStringRef p_input, MCStringLineEndingStyle p_to_style, MCStringLineEndingOptions p_options, MCStringRef& r_output, MCStringLineEndingStyle* r_initial_style);
```
This is much more verbose in use, but means that it is 100% clear in the points of use what the call is doing.

bwmilby commented:
I was actually thinking about a native line ending identifier for that enum which is good that I'm thinking along the same lines.  I'm still a little confused about OS X being identified as CR native though.

I wasn't certain what to replace "false" with for the MCStringLineEndingOptions parameter.  The compiler did not throw an error on it.

bwmilby committed:
Added MCStringLineEndingOptions enum and updated functions to use new values.  Moved native line ending handling to enum definition in the header file.

bwmilby committed:
Add test to verify that files retain line endings when edited in LiveCode.  Takes advantage of the fact that a UTF8 BOM is added to ensure that the file was actually saved.

montegoulding commented:
@livecode-vulcan review ok cdea5a0

@bwmilby in future could you try to make commit messages have a <80 char summary line so we can get the gist of the commit in one line history views. We have some guidelines for msg writing here: https://github.com/livecode/livecode/blob/develop/CONTRIBUTING.md

livecode-vulcan commented:
💙 review by @montegoulding ok cdea5a0a7d185a457cf0e4a290e74995861f55eb

bwmilby commented:
Roger.  Was only doing that for the initial PR commit.  After the initial PR/commit, does the bug need to be listed in each one or just the changes in that commit?

livecode-vulcan commented:
😎 test success cdea5a0a7d185a457cf0e4a290e74995861f55eb

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/674)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/690)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/664)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/607)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/610)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/668)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/688)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/673)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/690)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/672)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/683)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/705)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/669)
